massa-protocol-worker/src/protocol_network.rs

```diff
@@ -0,0 +1,426 @@
+    /// forward the block to the graph in that case.
+    ///
+    /// # Parameters:
+    /// - `from_node_id`: Node which sent us the information.
+    /// - `BlockId`: ID of the related operations we received.
+    /// - `operation_ids`: IDs of the operations contained by the block.
+    ///
+    /// # Result
+    /// return an error if stopping asking block failed. The error should be forwarded at the
+    /// root. todo: check if if make panic.
+    async fn on_block_operation_list_received(
+        &mut self,
+        from_node_id: NodeId,
+        block_id: BlockId,
+        operation_ids: OperationIds,
+    ) -> Result<(), ProtocolError> {
+        match self.block_wishlist.get(&block_id) {
+            Some(AskForBlocksInfo::Info) => {}
+            _ => return Ok(()),
+        }
+        let mut info = match self.checked_headers.get_mut(&block_id) {
+            Some(info) => info,
+            _ => {
+                warn!("Missing block header for {}", block_id);
+                return Ok(());
+            }
+        };
+        let mut total_hash: Vec<u8> = vec![];
+        operation_ids.iter().for_each(|op_id| {
+            let op_hash = op_id.hash().into_bytes();
+            total_hash.extend(op_hash);
+        });
+
+        // Check operation_list against expected operations hash from header.
+        if info.header.content.operation_merkle_root == Hash::compute_from(&total_hash) {
+            // Add the ops of info.
+            info.operations = Some(operation_ids.clone());
+
+            let missing_operations = operation_ids
+                .into_iter()
+                .filter(|op| {
+                    // Filter missing operations returning true if not found and compute
+                    // the sum of operation bytes size
+                    if let Some(operation_id) = self.checked_operations.get(&op.into_prefix()) {
+                        let op = self
+                            .storage
+                            .retrieve_operation(operation_id)
+                            .expect("unexpected checked operation not in storage");
+                        info.operations_size =
+                            info.operations_size.saturating_add(op.serialized_size());
+                        return false;
+                    }
+                    true
+                })
+                .collect();
+
+            if info.operations_size > self.config.max_serialized_operations_size_per_block {
+                let _ = self.ban_node(&from_node_id).await;
+                return Ok(());
+            }
+
+            // Update ask block
+            let mut set = Set::<BlockId>::with_capacity_and_hasher(1, BuildMap::default());
+            set.insert(block_id.clone());
+            self.stop_asking_blocks(set)?;
+
+            // Re-add to wishlist with new state.
+            self.block_wishlist
+                .insert(block_id, AskForBlocksInfo::Operations(missing_operations));
```

Comment:
It is probably possible to already have all operations for that block. In that case, we should already forward the block to the graph. Or I'm missing something else @gterzian ?

Reply:
Yes that's a good point, thanks.

Reply:
@gterzian was this solved ?